### PR TITLE
for #24177: sync tabs when home is shown

### PR DESCRIPTION
**Pull request by MatthewTighe**
Still have a few issues to work out, but I thought I'd open a draft PR in case there were any sweeping architectural changes requests. KDocs and tests to come once that stuff is settled.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt:57]: Just wanted to make sure that this change is intentional. `previewImageUrl` is a very distinct naming that follows the GV naming and is a preview (hero) image of the page and not an icon (hero image != icon). The preview (hero) image is parsed from the page's metadata (these are `<meta property="og:image">` tags).

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt:48]: I am starting to think we should just rip out the Recent Synced Tabs into its own component separate from `RecentTabs` and do the following:

```
Column {
  RecentTabs()
  Spacer()
  RecentSyncedTab()
}
```
This has the benefit of separating out the state and callback parameters of 2 very distinct components. I don't think Layout would be impeded either.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt:57]: Thanks for double-checking.  think I was incorrect in the initial implementation by naming it `previewImageUrl`. After some testing, it was looking like it pointed at a favicon and not a hero image. The `iconUrl` naming convention matches the underlying data that is stored as a `TabEntry` in `RemoteTabsStorage` in A-C.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt:57]: 👍

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt:48]: I separated them into two separate composables. Were you suggesting to also move the synced tab stuff out of `RecentTabController` and into its own handler?

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt:48]: I will leave that up to you. I am on the fence since there's a `RecentTab.SyncedTab` so maybe I'll draw your attention towards whether or not `SyncedTab` should belong in the sealed class of `RecentTab`. Before I think it made sense assuming we were appending to the `List<RecentTab>` and therefore determining whether or not we would render the component based on it, but if this is gonna be a totally unrelated state, maybe we should also move it out.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabs.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabs.kt:86]: Hm.. my original intention of having the 2 separate components was to dissolve having a @Composable that needed to take the combination of the `RecentSyncedTabState` and `List<RecentTab>` states and their respective callbacks since we can just pass these parameters to the components directly in `Content()` instead of having a wrapper component.

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabs.kt:86]: Ah, I see. I'll update that. I think it makes more sense to remove the `SyncedTab` from the `RecentTab` sealed class as well, so I'll do some further refactoring.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt:5]: Maybe move all of this into its own /home/recentsyncedtab/ package since we're breaking it out

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt:5]: I didn't at first because I wasn't sure where to put shared components between the two packages. I think it makes sense to continue sharing some of the composables in `RecentTabs`, since it seems likely that the design for the two cards will stay largely in sync. I can move the recent synced tab stuff into its own package and just import the recent tabs composables for now, but IMO that shows they are candidates for a more generic UI package in the future.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt:5]: Can you make a list of the reusable composable components? The one obvious one I saw in my brief scan is `RecentTabTitle`, which I think we should probably refactor and use our existing reusable `PrimaryText()` component and define a typography. This can be a filed as a followup along with the list of things you see that we can refactor. Alternatively, I would refactor them in a separate issue and land it before this PR.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:174]: I'll update it, but I'm curious as to why we'd like to. `Components.syncedTabsStorage` is `lazyMonitored` in case that hook is ever used again, and immediately initializes the underlying `Lazy<RemoteTabsStorage>`. The `syncedTabFeature` is also lazily instantiated in the `HomeFragment`.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentSyncedTabHomeFeature.kt:51]: Oh good catch, I'll remove that.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/interactor/RecentSyncedTabInteractor.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/interactor/RecentSyncedTabInteractor.kt:9]: Yeah will do! Been holding off on tests and kdocs until the shape of the code is in place.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:174]: Additionally, it doesn't look like there's a `Lazy<SyncedTabsStorage>`. Would it be best to add one?

**Comment by MatthewTighe** [comment]
@gabrielluong I think this should be ready for another round of review

**Comment by gabrielluong** [comment]
> @gabrielluong I think this should be ready for another round of review

Ok, I might not get to this today. I also get a ping on Slack from the GitHub app if you push the re-request review button.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:121]: Can we flip this so that we follow the order of the parameters in `PrimaryText()`?

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:138]: IIRC I think we can simplify this to `size(18.dp)` since the width and height are the same.

File: app/src/test/java/org/mozilla/fenix/home/SessionControlInteractorTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/home/SessionControlInteractorTest.kt:179]:
```suggestion
    fun `WHEN recent synced tab is clicked THEN the tab is handled`() {
```

File: app/src/test/java/org/mozilla/fenix/home/SessionControlInteractorTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/home/SessionControlInteractorTest.kt:187]:
```suggestion
    fun `WHEN recent synced tabs show all is clicked THEN show all synced tabs is handled`() {
```

File: app/src/test/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeatureTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeatureTest.kt:40]: Maybe go with `deviceAccessed1`, `deviceAccessed2`. Up to you

**Review by gabrielluong — CHANGES_REQUESTED** [review]
Looks pretty good so far.

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt:42]: I'd been thinking that `AppAction.Change` was intended to expose a way to manipulate the entire `AppState` in a single action, for setting up initial states in tests or otherwise. If that wasn't the intended use-case, then I can remove `recentSyncedTabState` from the action entirely.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt:31]: This is a super pedantic stance of mine, but I'd argue from the context of a client of the class, which is calling the constructor and arguably who the kdoc is for, has no knowledge of whether the constructor parameter is a property of the class or not since it's private. 

I'll switch it here, just wanted to drop my two cents for fun

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt:42]: We used to call `AppAction.Change` in `onStart` to populate the `HomeFragmentStore` with the cached state from various settings and storage. That was one of the reason it took so many params. We don't need to do this anymore since `AppStore` is global. Otherwise, it's current functionality is to change multiple parameters all at once easily in the BlocklistMiddle, but we will want to rethink what parameter needs to be exposed.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt:84]: Good catch, ty

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewHolder.kt:71]: Ah good call. I had initially thought to try and leave as much logic as possible out of the view holder, but I agree that could lead to to the above.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabHomeFeature.kt]: Dug in a bit and wrote a better description. Lemme know if it looks better

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt:42]: Makes sense! I don't think it will need to be included for now, then

**Comment by MatthewTighe** [comment]
@gabrielluong thanks as always for the thorough review. Think the only item left from this round is the performance benefits of using `Lazy` wrappers around some properties. It might be best to just go for it for now, since Jonathan is out. I think we want this in nightly as long as possible.

**Comment by MatthewTighe** [comment]
Looking more closely at the concerns about `Lazy` constructor values, I think we can ignore them for now.

- `accountManager` has become a param instead of a property. 
- `accountManager` and the `SyncedTabsStorage` are both only used to construct the underlying `SyncedTabsFeature`. The feature is instantiated lazily, so should not be constructed until a call to `start` triggered by lifecycle components.
- both components are also wrapped in a lazy delegate.

AFAIK, we should have lazy instantiation for every potentially expensive component. IMO, we should move forward with a merge and nightly release to give this as much exposure time as possible and revisit the `Lazy` stuff if there's a need down the line.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:55]: I would go one step further and just make this `RecentSyncedTab?` since we already checked the state prior to this point and avoid processing the state at this point.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:54]: Can we also add some @Preview for this component

File: app/src/test/java/org/mozilla/fenix/components/AppStoreTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/components/AppStoreTest.kt:255]: Probably don't need this.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:161]: I have a fixed for the color problems you were seeing in https://github.com/mozilla-mobile/fenix/issues/24653.

**Review by gabrielluong — APPROVED** [review]

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/view/RecentSyncedTab.kt:161]: TY!

**Pull request merged**
